File: server/src/collectors/unifiedBuses.js
/**
 * Unified Bus Network - fuses and deduplicates bus stops + terminals:
 *   - MLIT P11 bus stops (~200k points)
 *   - GTFS-JP nationwide aggregator (gtfs-data.jp)
 *   - busRoutes.js (OSM amenity=bus_station + curated highway terminals)
 *   - osmTransportBuses (always-on OSM transport layer for stops + terminals)
 *
 * Each feature is tagged with `kind: 'stop' | 'terminal'`.
 * Dedup key priority: GTFS-JP qualified stop_id -> name+coord-grid.
 */

import mlitP11BusStops from './mlitP11BusStops.js';
import gtfsJp from './gtfsJp.js';
import busRoutes from './busRoutes.js';
import osmTransportBuses from './osmTransportBuses.js';
import { createUnifiedCollector } from '../utils/unifiedCollectorTemplate.js';

export default createUnifiedCollector({
  sourceId: 'unified_buses',
  description: 'Deduplicated nationwide bus stops + terminals - fused MLIT P11 + GTFS-JP + OSM transport + curated highway terminals',
  upstreams: [
    { name: 'mlit-p11-bus-stops',  fn: mlitP11BusStops, kind: 'stop' },
    { name: 'gtfs-jp',             fn: gtfsJp,          kind: 'stop' },
    { name: 'bus-routes',          fn: busRoutes,       kind: 'terminal' },
    { name: 'osm-transport-buses', fn: osmTransportBuses /* upstream already tags kind */ },
  ],
  dedupeKeys: [
    (f) => {
      const id = f.properties?.stop_id;
      if (!id) return null;
      if (String(id).startsWith('GTFSJP_')) return id;
      return null;
    },
  ],
  dedupeOpts: { coordPrecision: 4 },
});

File: native/collectors/sources/unified_buses.c
/* collectors/transport/sources/unified_buses.c
 * P5 — port of server/src/collectors/unifiedBuses.js (createUnifiedCollector).
 * Fuses mlit-p11-bus-stops + gtfs-jp + bus-routes + osm-transport-buses (all
 * already-registered upstreams; first three tagged kind stop/stop/terminal,
 * the OSM layer self-tags kind), dedupes by GTFS-JP qualified stop_id then the
 * name+coord-grid fallback (coordPrecision 4). No postProcess. The `_meta`
 * envelope is dropped per RULE 8 (same as every other port). */
#include "../../source.h"
#include "../../lib/unified.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <string.h>
#include <ctype.h>

/* dedupeKeys[0]: id=stop_id; if(!id)null; String(id).startsWith("GTFSJP_")?id:null */
static const char *k_gtfsjp_stop_id(cJSON *f, char *b, size_t n) {
  cJSON *p = cJSON_GetObjectItem(f, "properties");
  cJSON *v = p ? cJSON_GetObjectItem(p, "stop_id") : NULL;
  char id[256] = {0};
  if (cJSON_IsString(v) && v->valuestring[0]) snprintf(id,sizeof id,"%s",v->valuestring);
  else if (cJSON_IsNumber(v)) snprintf(id,sizeof id,"%g",v->valuedouble);
  else return NULL;
  if (strncmp(id, "GTFSJP_", 7) != 0) return NULL;
  snprintf(b, n, "%s", id);
  return b;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  static const unified_upstream ups[] = {
    { "mlit-p11-bus-stops",  "stop"     },
    { "gtfs-jp",             "stop"     },
    { "bus-routes",          "terminal" },
    { "osm-transport-buses", NULL       },
  };
  static const unified_keyfn keys[] = { k_gtfsjp_stop_id };
  int n = unified_collect(ctx, sink, "unified-buses", ups, 4,
                          keys, 1, 4, NULL, NULL);
  return n >= 0 ? 0 : -1;
}

static const source_def unified_buses_def = {
  .id = "unified-buses", .collector = "transport",
  .name = "Unified Buses (fused)", .name_ja = "Unified Buses",
   .update_interval_sec = 86400, .run = run };
REGISTER_SOURCE(unified_buses_def)
